tests/test_coordinated_workers/test_worker.py

```diff
@@ -774,3 +775,119 @@ def test_worker_stop_all_services_if_not_ready(tmp_path):
     assert all(svc is ops.pebble.ServiceStatus.INACTIVE for svc in service_statuses), [
         stat.value for stat in service_statuses
     ]
+
+
+@patch("socket.getfqdn")
+def test_invalid_url(mock_socket):
+    # Test that when socket returns an invalid url as a Fully Qualified Domain Name,
+    #   ClusterRequirer.publish_unit_address raises a ValueError exception
+
+    # WHEN you define a properly configured charm
+    ctx = testing.Context(
+        MyCharm,
+        meta={
+            "name": "foo",
+            "requires": {"cluster": {"interface": "cluster"}},
+            "containers": {"foo": {"type": "oci-image"}},
+        },
+        config={"options": {"role-all": {"type": "boolean", "default": True}}},
+    )
+
+    # AND ClusterRequirer is passed an invalid url
+    mock_socket.return_value = "http://www.invalid-]url.com"
+
+    # IF the charm executes any event
+    # THEN the charm raises an error with the appropriate cause
+    with pytest.raises(UncaughtCharmError) as exc:
+        ctx.run(ctx.on.update_status(), testing.State(containers={testing.Container("foo")}))
+    assert isinstance(exc.value.__cause__, ValueError)
+
+
+@pytest.mark.parametrize(
+    "remote_databag, expected",
+    (
+        (
+            {
+                "charm_tracing_receivers": json.dumps({"url": "test-url.com"}),
+                "worker_config": json.dumps("test"),
+            },
+            {"url": "test-url.com"},
+        ),
+        (
+            {"charm_tracing_receivers": json.dumps(None), "worker_config": json.dumps("test")},
+            {},
+        ),
+    ),
+)
+def test_get_charm_tracing_receivers(remote_databag, expected):
+    ctx = testing.Context(
+        MyCharm,
+        meta={
+            "name": "foo",
+            "requires": {"cluster": {"interface": "cluster"}},
+            "containers": {"foo": {"type": "oci-image"}},
+        },
+        config={"options": {"role-all": {"type": "boolean", "default": True}}},
+    )
+    container = testing.Container(
+        "foo",
+        execs={testing.Exec(("update-ca-certificates", "--fresh"))},
+        can_connect=True,
+    )
+    relation = testing.Relation(
+        "cluster",
+        remote_app_data=remote_databag,
+    )
+    with ctx(
+        ctx.on.relation_changed(relation),
+        testing.State(containers={container}, relations={relation}),
+    ) as mgr:
+        charm = mgr.charm
+        mgr.run()
```

Comment:
I'm not sure in this test and the next you need to mgr.run() before your assertion. Could you check whether that is necessary?
The difference is subtle: mgr.run() emits the event, so the question is: does the return value of `get_charm_tracing_receivers` depend on the charm instance state? I hope and suspect that it shouldn't, so we can write the assertion without emitting the event first.

This also applies to the test below

Reply:
The test still pass with the assertions before mgr.run(). Should I execute mgr.run() anyways or remove it as it is unnecessary?

Reply:
remove it